tensorflow/core/kernels/data/dataset_test_base.h

```diff
@@ -0,0 +1,129 @@
+
+#ifndef TENSORFLOW_DATASET_TEST_BASE_H
+#define TENSORFLOW_DATASET_TEST_BASE_H
+
+#include <memory>
+#include <tuple>
+#include <vector>
+
+#include "tensorflow/core/framework/dataset.h"
+#include "tensorflow/core/framework/function.h"
+#include "tensorflow/core/framework/function_handle_cache.h"
+#include "tensorflow/core/framework/function_testlib.h"
+#include "tensorflow/core/framework/node_def_builder.h"
+#include "tensorflow/core/framework/partial_tensor_shape.h"
+#include "tensorflow/core/framework/variant.h"
+#include "tensorflow/core/framework/variant_tensor_data.h"
+#include "tensorflow/core/kernels/data/dataset_utils.h"
+#include "tensorflow/core/kernels/data/iterator_ops.h"
+#include "tensorflow/core/kernels/ops_testutil.h"
+#include "tensorflow/core/platform/test.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/util/ptr_util.h"
+
+namespace tensorflow {
+namespace data {
+
+class DatasetOpsTestBase : public ::testing::Test {
+ public:
+  DatasetOpsTestBase()
+      : device_(DeviceFactory::NewDevice("CPU", {}, "/job:a/replica:0/task:0")),
+        device_type_(DEVICE_CPU) {
+    CHECK(device_.get()) << "Could not create CPU device";
+    allocator_ = device_->GetAllocator(AllocatorAttributes());
+  }
+
+  ~DatasetOpsTestBase() {
+    gtl::STLDeleteElements(&tensors_);
+    gtl::STLDeleteElements(&managed_outputs_);
+  }
+
+  Status InitThreadPool(int thread_num);
+
+  Status InitFunctionLibraryRuntime(const std::vector<FunctionDef>& flib,
+                                    int cpu_num);
+
+  Status InitDatasetFromContext(OpKernelContext* context, int output_index,
+                                DatasetBase** dataset);
+
+  Tensor* GetOutput(OpKernelContext* context, int output_index, Status* status);
+
+  Status InitOpKernelContext(OpKernel* kernel,
+                             gtl::InlinedVector<TensorValue, 4>* inputs,
+                             std::unique_ptr<OpKernelContext>* context);
+
+  Status RunOpKernel(OpKernel* op_kernel, OpKernelContext* context);
+
+  void InitSerializationContext(
+      std::unique_ptr<SerializationContext>& serialization_context_);
+
+  Status MakeOpKernel(NodeDef node_def, std::unique_ptr<OpKernel>& kernel);
+
+  template <typename T>
+  void AddDatasetInputFromArray(gtl::InlinedVector<TensorValue, 4>* inputs,
+                                DataTypeVector input_types,
+                                const TensorShape& shape,
+                                const gtl::ArraySlice<T>& data) {
+    Tensor* input =
+        AddDatasetInput(inputs, input_types, DataTypeToEnum<T>::v(), shape);
+    test::FillValues<T>(input, data);
+  }
+
+  Tensor* AddDatasetInput(gtl::InlinedVector<TensorValue, 4>* inputs,
+                          DataTypeVector input_types, DataType dtype,
+                          const TensorShape& shape);
+
+  Status MakeDataset(OpKernel* kernel, OpKernelContext* context);
```

Comment:
this method should return `DatasetBase*` through an output argument instead of updating internal state